### PR TITLE
NFT analyzer

### PR DESCRIPTION
**Pull request by pro-wh**
an analyzer for NFTs. the term "NFT" is proposed to refer to specific instances managed by a contract.

this code highly experimental and missing some important things.

- NFT analysis jobs involve downloading a "metadata" file from a URI given by a smart contract. currently those jobs may fail. we do not have a system for scheduling those jobs to be retried
- a contract may point to a server that is arbitrarily slow. currently we set a timeout of 30 seconds for downloading this file. but yeah, anyone could make the analyzer as slow as 30s/batch
- we have no way to start on more jobs if only a few downloads are straggling.
- we have no policy for judging what errors to treat as permanent, when it comes to downloading this metadata file

conflicts:

- ~~there's ongoing work #511 to build a superclass for analyzers like this. can we use that here? in anticipation, I have not written much to go around the actual download logic, e.g. it's a single-threaded download loop in this PR~~ integrated, thanks!
- ~~I think I saw there's something in the reckoning fixes #512 for unifying certain analysis.something and chain.something tables. might the principle leading to that apply here too?~~ unified woo

### CODE REVIEW

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [inline analyzer/runtime/evm/erc721.go:109]: we're starting to bleed out of evm-related things in this file :skull:

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [inline analyzer/runtime/evm/erc721.go:111]: big implications here, from the possibility that the metadata request goes to something other than our own node

1. the server could be slow. even maliciously slow. there's a 30s timeout, but like that means someone could make every batch take 30 seconds
2. some jobs can now fail without it being a horrible "our node is down and nothing else will ever work" situation, as was the case before. we'll want a way to continue with other work
3. some failures probably will be intermittent. after all, these requests will go over the internet, right? we'll want to hang on to a job so we can retry it later
4. oh and "later" will best be one of those backoff systems
5. all that, while we shouldn't assume that the list of all NFTs fits in memory

**Comment by pro-wh** [comment]
opening this for "review," although it might be easier not to merge this for quite some time. there's no API for the data this collects, and the analyzer is very un-robust, and it's missing some features that will affect the tables added here

File: analyzer/runtime/evm/erc721.go
Comment (mitjat) [reply analyzer/runtime/evm/erc721.go:111]: I'm trying to foresee what the architecture will be like, longer term. I expect there will be several async phases:
- Token id -> token metadata URL (by issuing an EVM query to the runtime).
- Metadata fetching (URL -> JSON) 
- Metadata analysis. This is almost trivial; just extracting the relevant field(s) from the metadata JSON.
- Image fetching (URL -> blob)

The reason I think it probably makes sense for each of these to be async from other is that the URL fetching is a) fallible and b) a reusable component.

Phases 2 and 4 are pretty much identical when it comes to requirements: URLs are untrusted, the fetched resource might become stale (I'm referring to non-constant NFTs here), fetching might need to be retried multiple times with backoff. So one possible solution would be to have a "resource fetching" analyzer: Stages 1 and 3 give jobs to it, and tag the job somehow to signify the table into which the result should be output. So the analyzer has the understanding of those tags built in. Or, it could be two separate analyzers, with a largely shared codebase, but each instantiated with a separate `dataInserter` struct (much like `processor` for blockAnalyzers) that knows what to do with the fetched data.

I don't think we need all of the above in this PR (or that it'd be an appropriate amount of new code for one PR, at that :)), but it would probably be helpful to align things with the architecture we want to move towards. For example, if we go with what I outlined above, I think it would make sense to add a minimal `resourceFetchAnalyzer` now. No retries, cache expiration, etc, just a naive one-off fetch.

Depending on how much our thought (mis)align, this might be worth a short doc, actually. I think neither the structure of the NFT analyzer(s) nor the structure of a resource fetcher/cache is clear yet.

**Review by mitjat — COMMENTED** [review]
Lots of work! Thank you.
Also, good points about future work in the PR description. I started a thread to hear your thinking on how we could address those, and to add thoughts of my own.

File: analyzer/runtime/evm/erc721.go
Comment (mitjat) [reply analyzer/runtime/evm/erc721.go:111]: Also wanted to point out that only phase 1 is EVM specific, which would be nice to reflect in analyzer names. (Just `nft`, no `evm_nft`)

File: analyzer/drawerfullofprotocols/drawer_full_of_protocols.go
Comment (mitjat) [reply analyzer/drawerfullofprotocols/drawer_full_of_protocols.go:1]: :raised_eyebrow:  how did you get into my house?

File: storage/migrations/18_evm_nfts.up.sql
Comment (pro-wh) [reply storage/migrations/18_evm_nfts.up.sql:13]: ok I'll be trying that

File: storage/migrations/18_evm_nfts.up.sql
Comment (pro-wh) [reply storage/migrations/18_evm_nfts.up.sql:10]: just the url

can we put that image fetch in the frontend at first?

File: cmd/analyzer/analyzer.go
Comment (pro-wh) [reply cmd/analyzer/analyzer.go:217]: ok

File: analyzer/drawerfullofprotocols/drawer_full_of_protocols.go
Comment (pro-wh) [reply analyzer/drawerfullofprotocols/drawer_full_of_protocols.go:1]: huh yeah, makes sense not to lump them together so much. leme put these inside the one analyzer where they're used for now.

File: analyzer/ipfsc/gateway.go
Comment (pro-wh) [reply analyzer/ipfsc/gateway.go:41]: the problem with doing that here is checking the response status code is a step. writing out the separate error return on the last fallible step is in the same vein as using the trailing commas in multiline lists

File: analyzer/queries/queries.go
Comment (pro-wh) [reply analyzer/queries/queries.go:621]: at this point I want to get this in and work on retry logic later

File: storage/migrations/18_evm_nfts.up.sql
Comment (pro-wh) [reply storage/migrations/18_evm_nfts.up.sql]: combined

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go:111]: oh I forgot to remove the "evm" part

File: analyzer/neededforhttp/needed_for_http.go
Comment (pro-wh) [reply analyzer/neededforhttp/needed_for_http.go]: the chunk of opinions encoded in this file are common to (i) that package formerly called pubclient, for making requests to an externally provided URL without SSRF; and (ii) accessing IPFS content over an IPFS HTTP gateway. unfortunately, (i) and (ii) can't be in common because it's intended that we use a local IPFS node, and SSRF prevention would prevent us from connecting to a local service. so (i) and (ii) are separate, and this code is not particularly closer to either of them, so it's in a third package now :smiling_face_with_tear:

File: analyzer/ipfsc/gateway.go
Comment (pro-wh) [reply analyzer/ipfsc/gateway.go]: renamed

File: analyzer/drawerfullofprotocols/drawer_full_of_protocols.go
Comment (pro-wh) [reply analyzer/drawerfullofprotocols/drawer_full_of_protocols.go]: renamed

File: analyzer/ipfsc/gateway.go
Comment (pro-wh) [reply analyzer/ipfsc/gateway.go]: okay something unrelated made me have to change the way this goes (omg was that a pun). the `bodyclose` lint can't detect that the resp.Body lives on if ResponseOK extracts it for us. so now ResponseOK no longer 'returns' anything (ok it still returns an error :roll_eyes:) and we do `return resp.Body, nil` instead

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go:112]: okay counter proposal: as this analyzer still has the EVM part, let's leave the "evm" in the name. we can separate out the network download later, along with tackling server errors

File: analyzer/evmnfts/evm_nfts.go
Comment (mitjat) [inline analyzer/evmnfts/evm_nfts.go:62]: Why not use common.BigInt here? The main point of introducing our own BigInt was that it interops more smoothly with pgx and other (de)serializations. Also, doesn't have funky pointer semantics.

File: analyzer/runtime/evm/erc721.go
Comment (mitjat) [inline analyzer/runtime/evm/erc721.go:93]: Nit: It _is_ a little hard to read. Can we do an early exit instead? `if !supportMetadata { return ... }`

Actually, looks like we could do early return on several more ifs below, and have this function be pretty much linear. That would help readability a fair bit.

File: analyzer/runtime/evm/erc721.go
Comment (mitjat) [inline analyzer/runtime/evm/erc721.go:99]: Nit: Can you add "deterministically" to the log message inside this helper func, and/or some other reassuring phrase? Right now if I saw that log, I'd be a little alarmed and felt like I needed to investigate.

File: analyzer/runtime/evm/erc721.go
Comment (mitjat) [reply analyzer/runtime/evm/erc721.go:112]: sgtm for the rename.

For the network download: if I understand the code correctly, any failed downloads currently will just cause the token to stay in the analyzer queue, so if we add a more complex `downloadAnalyzer` later, it will "just work" in that we won't need any complex db migrations to construct a work queue for it. If so, SGTM. Can you create a ticket in that case, please?

File: analyzer/evmnfts/evm_nfts.go
Comment (mitjat) [inline analyzer/evmnfts/evm_nfts.go:117]: this can probably be a single query now?

File: storage/migrations/18_evm_nfts.up.sql
Comment (mitjat) [reply storage/migrations/18_evm_nfts.up.sql]: Yeah IIRC @lukaw3d said there's no csrf in place because the explorer is read-only. So for the short term, letting them fetch directly from untrusted sources should be fine.

File: cmd/analyzer/analyzer.go
Comment (mitjat) [reply cmd/analyzer/analyzer.go]: not implemented, I think

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (mitjat) [inline analyzer/evmnfts/httpmisc/http_misc.go:38]: Here and below: Let's include the HTTP status code in the error.

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (mitjat) [inline analyzer/evmnfts/httpmisc/http_misc.go:46]: What is the point of NotOKError? It's an "enum" with a single value, it's produced seemingly inconsistently (I don't understand why the ability to read the response body of e.g. a HTTP 404 would differentiate between a `NotOKError` and a `fmt.Errorf`) and nobody "reads" it, i.e. checks for its presence.
Can we get rid of it and reintroduce when/if requirements are clearer?

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (mitjat) [inline analyzer/evmnfts/httpmisc/http_misc.go:36]: I don't understand how this is different from `StatusCode != 200`; the handler seems identical, except for the subtle difference in producing the `NotOKError` wrapper or not.

File: analyzer/neededforhttp/needed_for_http.go
Comment (mitjat) [reply analyzer/neededforhttp/needed_for_http.go]: Hrm yeah I wanted these funcs to be embedded in `multiproto`, but that would cause circular dependencies with (i) and (ii) :/
Can you add top-level descriptions to these new modules? They're kinda vaguely/broadly named, and at the same time contain so little functionality that it's hard to see at a glance what their intended scope is.

File: analyzer/queries/queries.go
Comment (mitjat) [reply analyzer/queries/queries.go]: OK. To me, it looks easier (and fast enough) to change it now that to deal with migrations in the future. But whatever is your preference.

**Review by mitjat — APPROVED** [review]
Thank you Warren!

I only have small comments left at this point, so LGTM pending resolved comment threads.

File: storage/migrations/18_evm_nfts.up.sql
Comment (lukaw3d) [reply storage/migrations/18_evm_nfts.up.sql]: Yep yep, we don't have CSP yet
(also no CSRF, but that's different)

File: cmd/analyzer/analyzer.go
Comment (pro-wh) [reply cmd/analyzer/analyzer.go]: woops yeah, gotta add that

File: analyzer/neededforhttp/needed_for_http.go
Comment (pro-wh) [reply analyzer/neededforhttp/needed_for_http.go]: ok adding comments

File: analyzer/queries/queries.go
Comment (pro-wh) [reply analyzer/queries/queries.go]: let me file something for that

Comment (pro-wh) [reply analyzer/queries/queries.go]: #532

File: analyzer/evmnfts/evm_nfts.go
Comment (pro-wh) [reply analyzer/evmnfts/evm_nfts.go:62]: lately I've been using a lot more go-ethereum types, which it uses *big.Int for its large integers. common.BigInt does come in handy when scanning from the database though.

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go:93]: early exits are more suitable for "something went wrong, bail." this is a section where we progressively obtain more info if it does support a certain interface. not supporting that interface is fine, and we may theoretically do other things to get information. following these guidelines for when to use early exit and when to do something inside an `if` block helps make code more conveniently composable

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go:99]: ya adding

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go:93]: I'll break this up into other methods tho. would be good to early-exit the metadata part if calling tokenURI fails for example

File: analyzer/evmnfts/evm_nfts.go
Comment (pro-wh) [reply analyzer/evmnfts/evm_nfts.go:117]: oh yeah

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go:112]: > any failed downloads currently will just cause the token to stay in the analyzer queue

no, there's more fiddliness to it. right now:

"DONE" means download job is successfully completed, possibly with no or partial info collected; "BREAK" means something really bad is happening and abort the whole batch. currently there's no third option for "just this one has a problem"

- problem checking for ERC721Metadata ->
  - deterministic -> ERC165 says this means it's not supported. (continue as if not supported)
  - nondeterministic -> something's wrong with our access to the network. BREAK
- doesn't support ERC721Metadata -> that's fine, no data from this source. DONE
- supports ERC721Metadata -> (query tokenURI)
  - problem querying tokenURI ->
    - deterministic -> contract's fault. DONE
    - nondeterministic -> something's wrong with our access to the network. BREAK
  - got token URI -> (save token URI and "accessed" date)
    - URI empty or malformed etc. -> contract's fault. DONE
    - URI unsupported scheme -> uh... just give up. DONE
    - URI parseable and supported -> (try to download)
      - just about anything goes wrong -> not necessarily NFT's fault, but also don't want to let this starve the queue. DONE :arrow_left: **this thing**
      - download OK -> (try to parse the info JSON doc)
        - problem parsing JSON -> NFT's fault. DONE
        - JSON OK -> close download stream
          - problem closing download stream -> Go runtime integrity problem?? BREAK
          - close stream OK -> (save name + description + image URL) DONE

so the main thing labeled "this thing" above, that's where currently we treat that as "DONE" and it's arguably bad. this PR has the beginnings of some things that'll help us tell if it's our fault or the NFT's fault, but it's not in use yet. we'll probably have to do a migration to repopulate the queue with jobs that ended up this way. a superset of those jobs, really

but still yeah, filed in #532.

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (pro-wh) [reply analyzer/evmnfts/httpmisc/http_misc.go:38]: ok

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (pro-wh) [reply analyzer/evmnfts/httpmisc/http_misc.go:46]: meant for a future update to separate whether to retry. statuses 5xx indicate server error, which would be a good candidate for retry.

4xx indicates the server isn't satisfied with our request, so it's probably better to give up. Come to think of it, 429 might be better in the other category.

3xx is supposed to be a redirect, but the HTTP client ought to handle that for us. if for whatever reason it didn't, then it's not like we can handle it out here, so it's probably better to give up.

2xx other than 200 ("OK") indicates some kind of successful request but where it doesn't actually give us a file. we're always looking for a file, so it's probably better to give up.

1xx is some message used in the HTTP protocol but aren't formally the response. I suspect the HTTP client handles these internally, but if it somehow gets out here, then we can't handle it, so it's probably better to give up.

These "give up" cases are lumped into a `NotOKError`

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (pro-wh) [reply analyzer/evmnfts/httpmisc/http_misc.go:36]: answering here https://github.com/oasisprotocol/nexus/pull/506#discussion_r1339330308

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go]: added

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go]: refactored

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go]: removed

File: cmd/analyzer/analyzer.go
Comment (pro-wh) [reply cmd/analyzer/analyzer.go]: added

File: analyzer/drawerfullofprotocols/drawer_full_of_protocols.go
Comment (pro-wh) [reply analyzer/drawerfullofprotocols/drawer_full_of_protocols.go]: moved

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (pro-wh) [reply analyzer/evmnfts/httpmisc/http_misc.go]: included

File: analyzer/evmnfts/evm_nfts.go
Comment (mitjat) [reply analyzer/evmnfts/evm_nfts.go:62]: Yeah looking at the DB code was what prompted this comment.
I think this is the right layer to convert from whatever types go-ethereum gives us to the types we prefer to use internally, if the two are different.
Obviously a minor comment though, your call in the end, I won't cry either way.

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (mitjat) [reply analyzer/evmnfts/httpmisc/http_misc.go:49]: Alright thanks for the explanation. (And great catch with `HTTP 429 Too Many Requests`! I agree. Wanna encode that now?)
In that case, two small followups:
- Let's rename `NotOKError` please. The name misled me when reading because I think of "error" and "not ok" as pretty synonymous. Maybe `NonRetriableError`? Or we could reuse the name from the RPC cache: `DeterministicError`. Or similar.
- In the non-5xx case, even if we fail to `Close()` the body, I'd say that's still a `DeterministicError`. The HTTP status code is all we need to determine (heh) that. We can just silently log the `Close()` problem maybe? Or even ignore it altogether, assuming some resource cleanup happens behind the scenes.

File: analyzer/runtime/evm/erc721.go
Comment (pro-wh) [reply analyzer/runtime/evm/erc721.go]: :skull: resolving this to focus on active discussions

File: storage/migrations/18_evm_nfts.up.sql
Comment (pro-wh) [reply storage/migrations/18_evm_nfts.up.sql]: resolving this to focus on active discussions :skull:

File: analyzer/drawerfullofprotocols/drawer_full_of_protocols.go
Comment (pro-wh) [reply analyzer/drawerfullofprotocols/drawer_full_of_protocols.go]: haven't moved the other non-analyzer subdirectories though. so it continues :skull:

File: analyzer/evmnfts/evm_nfts.go
Comment (pro-wh) [reply analyzer/evmnfts/evm_nfts.go:62]: should we just say it? do you want us to prefer using common.BigInt and converting to *big.Int locally where needed?

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (pro-wh) [reply analyzer/evmnfts/httpmisc/http_misc.go:39]: closing this, let's discuss in the other conversation :skull:

File: analyzer/evmnfts/evm_nfts.go
Comment (mitjat) [reply analyzer/evmnfts/evm_nfts.go:62]: oh. Not sure what "should we just say it" means exactly, but yes, I do prefer what you describe. I think it makes for more consistent code and smoother interop between various parts of the indexer. (EVM work that often needs *big.Int because of go-ethereum is an unfortunate exception.)
So inasmuch as I have decision power, I choose to wield it in that direction :)

Comment (pro-wh) [reply analyzer/evmnfts/evm_nfts.go:62]: k going with common.BigInt here :tipping_hand_person:

File: analyzer/evmnfts/httpmisc/http_misc.go
Comment (pro-wh) [reply analyzer/evmnfts/httpmisc/http_misc.go:49]: changed to ResourceError, to convey that the server successfully answered about the Resource that the URL Locates, but we still find it unsatisfactory

**Pull request merged**
